prow/cmd/deck/main.go

```diff
@@ -408,6 +431,171 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
 	}
 }
 
+// handleRequestProwJobViews pre-renders a Spyglass page for a Prowjob source
+// A valid Prow job view url takes the form:
+//
+// /view/prowjob/<jobname>/<buildID>
+//
+// Example:
+// - /view/prowjob/echo-test/1021530234601607168
+func handleRequestProwJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		src := strings.TrimPrefix(r.URL.Path, "/view/")
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return Prowjob view page: ", elapsed)
+	}
+}
+
+// handleRequestGCSJobViews pre-renders a Spyglass page for a GCS source
+// A valid job GCS Job view url takes the form:
+//
+// /view/gcs/<bucketname>/<jobprefix>
+//
+// Example:
+// - /view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/121
+func handleRequestGCSJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		srcData := strings.TrimPrefix(r.URL.Path, "/view/gcs/")
+		src := fmt.Sprintf("gs://%s", srcData)
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return GCS view page: ", elapsed)
+	}
+}
+
+// handleRequestJobViews handles requests to get all available artifact views for a given job via
+// a general src parameter, which can contain any string used to obtain job artifacts
+// A valid general job view url takes the form:
+//
+// /view/?src=<URI-encoded-source-string>
+//
+// Example:
+// - /view/?src=gs:%2F%2Fkubernetes-jenkins%2Fpr-logs%2Fpull%2Fkubeflow_kubeflow%2F1195%2Fkubeflow-presubmit%2F2558
+func handleRequestJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		src := r.URL.Query().Get("src")
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return view page: ", elapsed)
+	}
+}
+
+// renderSpyglass returns a pre-rendered Spyglass page from the given source string
+func renderSpyglass(sg *spyglass.Spyglass, ca *config.Agent, src string) string {
+	artifactNames, err := sg.ListArtifacts(src)
+	if err != nil {
+		logrus.WithError(err).Error("Error listing artifacts")
+	}
+
+	viewerCache := map[string][]string{}
+	viewersRegistry := ca.Config().Deck.Spyglass.Viewers
+
+	for re, viewerNames := range viewersRegistry {
+		matches := []string{}
+		r, err := regexp.Compile(re)
+		if err != nil {
+			logrus.WithError(err).Error("Regexp failed to compile.")
```

Comment:
Can we configure this at config-load time?

Reply:
Definitely can, wasnt sure if we wanted to block loading config on an invalid regex. Anyone else have opinions on this?

Reply:
@cjwagner @krzyzacy